README.md
<p>
  <img height="128" src="https://www.checklyhq.com/images/footer-logo.svg" align="right" />
  <h1>Checkly Pulumi Provider</h1>
</p>

![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/checkly/pulumi-checkly?label=Version)

The Checkly Pulumi provider enables you to create and configure Checkly resources using your favourite programming language.
Note that this project is in its early stages and breaking changes could happen.

## Installation

1. To use this package, please [install the Pulumi CLI first](https://www.pulumi.com/docs/get-started/install/).
2. This package is only available for JavaScript and TypeScript but support for other languages will be available soon.

### Node.js (JavaScript/TypeScript)

To use from JavaScript or TypeScript in Node.js, install using either `npm` or `yarn`:

```bash
$ npm install @checkly/pulumi
$ yarn add @checkly/pulumi
```

### Python, Go & .NET

*TBA*

## Authentication

The Pulumi Checkly Provider needs to be configured with a Checkly `API Key` and `Account ID` before it can be used to create resources.

> If you don't have an `API Key`, you can create one [here](https://app.checklyhq.com/settings/user/api-keys).

Once you generated the `API Key` there are two ways to communicate your authorization tokens to Pulumi:

1. Set the environment variables `CHECKLY_API_KEY` and `CHECKLY_ACCOUNT_ID`:
    ```bash
    $ export CHECKLY_API_KEY=cu_xxx
    $ export CHECKLY_ACCOUNT_ID=xxx
    ```

2. Set them using `pulumi config` command, if you prefer that they be stored alongside your Pulumi stack for easy multi-user access:
    ```bash
    $ pulumi config set checkly:apiKey cu_xxx --secret
    $ pulumi config set checkly:accountId xxx
    ```

> Remember to pass `--secret` when setting `checkly:apiKey` so it is properly encrypted.

## Getting Started

1. Open your terminal and run `$ pulumi new javascript` to create a new Pulumi project with the `javascript` template.
1. Install the Checkly Pulumi provider using npm: `$ npm i @checkly/pulumi`.
1. Look for `index.js` file in the root of your project and replace content with the following code:

    ```javascript
    const checkly = require("@checkly/pulumi")

    new checkly.Check("api-check", {
      type: "API",
      name: "Public SpaceX API",
      activated: true,
      frequency: 10,
      locations: ["us-east-1"],
      request: {
        method: "GET",
        url: "https://api.spacexdata.com/v3",
        assertions: [
          {
            source: 'STATUS_CODE',
            comparison: 'EQUALS',
            target: 200
          },
          {
            source: 'JSON_BODY',
            property: '$.project_name',
            comparison: 'EQUALS',
            target: 'SpaceX-API'
          }
        ]
      }
    })

    new checkly.Check("browser-check", {
      type: "BROWSER",
      name: "Google.com Playwright check",
      activated: true,
      frequency: 10,
      locations: ["us-east-1"],
      script: `const { chromium } = require('playwright')

    async function run () {
      const browser = await chromium.launch()
      const page = await browser.newPage()

      const response = await page.goto('https://google.com')

      if (response.status() > 399) {
        throw new Error('Failed with response code \${response.status()}')
      }

      await page.screenshot({ path: 'screenshot.jpg' })

      await page.close()
      await browser.close()
    }

    run()`
    })
    ```
1. Setup you Checkly API Key and Account id:
    ```bash
    $ pulumi config set checkly:apiKey cu_xxx --secret
    $ pulumi config set checkly:accountId xxx
    ```
1. You are ready to go, run `$ pulumi up` to deploy your stack 🚀

## Examples

Find working JavaScript and TypeScript code samples in the [`./examples`](https://github.com/checkly/pulumi-checkly/tree/main/examples) directory.

## Learn More
For documentation and example usage see:
1. [Checkly's documentation](https://www.checklyhq.com/docs/integrations/pulumi/).
2. [The official provider documentation](https://www.pulumi.com/registry/packages/checkly/api-docs/)
3. [Working Examples](https://github.com/checkly/pulumi-checkly/examples).

## Questions
For questions and support please open a new  [discussion](https://github.com/checkly/pulumi-checkly/discussions). The issue list of this repo is exclusively for bug reports and feature/docs requests.

## Issues
Please make sure to respect issue requirements and choose the proper [issue template](https://github.com/checkly/pulumi-checkly/issues/new/choose) when opening an issue. Issues not conforming to the guidelines may be closed.

## Contribution
Please make sure to read the [Contributing Guide](https://github.com/checkly/pulumi-checkly/blob/main/CONTRIBUTING.md) before making a pull request.

## License

[MIT](https://github.com/checkly/pulumi-checkly/blob/main/LICENSE)

<br>
<p align="center">
  <a href="https://checklyhq.com?utm_source=github&utm_medium=sponsor-logo-github&utm_campaign=headless-recorder" target="_blank">
  <img width="100px" src="https://www.checklyhq.com/images/text_racoon_logo.svg" alt="Checkly" />
  </a>
  <br />
  <i><sub>Delightful Active Monitoring for Developers</sub></i>
  <br>
  <b><sub>From Checkly with ♥️</sub></b>
<p>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<br>

## [v1.1.3](https://github.com/checkly/pulumi-checkly/releases/tag/v1.1.3) - 2022-12-02

### Added
- Add missing display name to json schem [#35](https://github.com/checkly/pulumi-checkly/pull/35)

### Changed
- Bump github.com/checkly/terraform-provider-checkly to [`v1.6.3`](https://github.com/checkly/terraform-provider-checkly/blob/main/CHANGELOG.md#v163---2022-10-26) [#36](https://github.com/checkly/pulumi-checkly/pull/36)

## [v1.1.2](https://github.com/checkly/pulumi-checkly/releases/tag/v1.1.2) - 2022-08-03

### Changed
- Bump github.com/checkly/terraform-provider-checkly to [`v1.6.2`](https://github.com/checkly/terraform-provider-checkly/blob/main/CHANGELOG.md#v162---2022-08-02) [#f04b929](https://github.com/checkly/pulumi-checkly/commit/f04b929)

## [v1.1.1](https://github.com/checkly/pulumi-checkly/releases/tag/v1.1.1) - 2022-07-06

### Changed
- Bump github.com/checkly/terraform-provider-checkly to [`v1.6.1`](https://github.com/checkly/terraform-provider-checkly/blob/main/CHANGELOG.md#v161---2022-07-06) [#c200aa1](https://github.com/checkly/pulumi-checkly/commit/c200aa1)

## [v1.1.0](https://github.com/checkly/pulumi-checkly/releases/tag/v1.1.0) - 2022-06-23

### Added
- Support for Private Locations [#24](https://github.com/checkly/pulumi-checkly/issues/24)

### Changed
- Bump github.com/checkly/terraform-provider-checkly to [`v1.6.0`](https://github.com/checkly/terraform-provider-checkly/blob/main/CHANGELOG.md#v160---2022-06-21)

## [v1.0.1](https://github.com/checkly/pulumi-checkly/releases/tag/v1.0.1) - 2022-06-21

### Fixed
- Update lint errors on docs [#7144661](https://github.com/checkly/pulumi-checkly/commit/714466136478e4e42c2e4664cf5e82ee59e189c6)

## [v1.0.0](https://github.com/checkly/pulumi-checkly/releases/tag/v1.0.0) - 2022-06-16

> Hello World! 👋


examples/js/index.js
const fs = require('fs')
const checkly = require("@checkly/pulumi")

// We store code snippets separately from the Checkly entities.
const snippetScript = fs.readFileSync("./scripts/snippet-script.js", "utf-8")
const playwrightScript = fs.readFileSync("./scripts/playwright-script.js", "utf-8")

// Some variables we can reuse in checks and check groups
const locations = ['eu-west-1', 'us-west-2']
const tags = ['pulumi']

// Alert channels are defined up front so we can add them to checks and check groups later.
const emailChannel = new checkly.AlertChannel("my-email-channel", {
  email: {
    address: 'your@email.com'
  }
})

const slackChannel = new checkly.AlertChannel("my-slack-channel", {
  slack: {
    url: 'https://hooks.slack.com/services/<REPLACE_WITH_ACTUAL_SLACK_HOOK>',
    channel: '#alerts',
  }
})

// We create a group, add the tags, locations and alert channels.
const group = new checkly.CheckGroup("my-group", {
  name: 'Check Group #1',
  activated: true,
  concurrency: 1,
  apiCheckDefaults: {},
  locations,
  tags,
  useGlobalAlertSettings: true,
  alertChannelSubscriptions: [
    {
      activated: true,
      channelId: slackChannel.id.apply(id => parseInt(id)),
    },
    {
      activated: true,
      channelId: emailChannel.id.apply(id => parseInt(id)),
    }
  ]
})

// Our first Browser check is added to group created above and uses the Playwright script.
new checkly.Check("my-browser-check", {
  name: "Google.com Playwright check",
  activated: true,
  frequency: 10,
  type: "BROWSER",
  groupId: group.id.apply(id => parseInt(id)),
  script: playwrightScript,
  locations,
  tags
})

const setupSnippet = new checkly.Snippet('my-snippet', {
  name: 'set-header',
  script: snippetScript
})

// Our first API check uses the setup script defined above and has two assertions.
new checkly.Check("my-api-check", {
  name: "Public SpaceX API",
  activated: true,
  frequency: 10,
  type: "API",
  locations,
  tags,
  setupSnippetId: setupSnippet.id.apply(id => parseInt(id)),
  degradedResponseTime: 5000,
  maxResponseTime: 15000,
  request: {
    method: "GET",
    url: "https://api.spacexdata.com/v3",
    assertions: [
      {
        source: 'STATUS_CODE',
        comparison: 'EQUALS',
        target: 200
      },
      {
        source: 'JSON_BODY',
        property: '$.project_name',
        comparison: 'EQUALS',
        target: 'SpaceX-API'
      }
    ]
  },
  useGlobalAlertSettings: true,
  alertChannelSubscriptions: [
    {
      activated: true,
      channelId: emailChannel.id.apply(id => parseInt(id)),
    }
  ]
})

// This maintenance window runs for 1 hour after midnight UTC and repeats daily for two months.
// All the checks with the "tags" array are included in this window.
new checkly.MaintenanceWindow('my-maintenance-window', {
  name: "Daily maintenance",
  startsAt: '2022-03-01T00:00:00.000Z',
  endsAt: '2022-03-01T01:00:00.000Z',
  repeatEndsAt: '2022-05-02T00:00:00.000Z',
  repeatInterval: 1,
  repeatUnit: 'DAY',
  tags
})

// The public dashboard shows all the checks tags with the "tags" array
new checkly.Dashboard('my-dashboard', {
  customUrl: "my-pulumi-checkly-dash",
  customDomain: "",
  header: "Checks created with Pulumi",
  hideTags: false,
  logo: "",
  paginate: false,
  paginationRate: 30,
  refreshRate: 300,
  tags
})

// Environment variables can be used in API request URLs, headers, query parameters etc.
// Use variables in API checks by using the {{MY_VAR}} notation.
// Use variables in Browser checks with the process.env.MY_VAR notation
new checkly.EnvironmentVariable('my-env-var', {
  key: 'API_URL',
  value: 'https://localhost:3000',
})